README.md
# Kaal  
![](https://img.shields.io/youtube/views/_bkIw7igeJM?label=Views&style=for-the-badge) [Video Link](https://www.youtube.com/watch?v=_bkIw7igeJM)
  
**🤖 [Invite the Bot to your server](https://discord.com/oauth2/authorize?client_id=763677336813436950&permissions=8&scope=bot)**
  
![](https://img.shields.io/badge/commit%20activity-+70-blue) ![](https://github.com/MLH-Fellowship/0.2.1-fellowbook) ![](https://img.shields.io/badge/build-passing-brightgreen) ![](https://img.shields.io/badge/contributors-only%203-orange) ![](https://img.shields.io/badge/version-1.0.0-yellow) ![](https://img.shields.io/badge/learned%20a%20lot-yes-blue)  
  
![Banner](https://github.com/MLH-Fellowship/pod_1.0.2_moropy/blob/master/branding/github_banner.png)  

<p align="center">
    October 12th, 2020 - <a href='https://fellowship.mlh.io/'>MLH Fellowship</a> Orientation Hackathon Project<br>
    by <a href='https://github.com/zerefwayne'>Aayush Joglekar</a>, <a href='https://github.com/pandafy'>Gagan Deep</a> and <a href='https://github.com/rish07'>Rishi Raj Singh</a>.
</p>

## 💁‍♂️ What is the project?

**Kaal is a productivity suite with a beautiful CLI ✨ and Discord Bot 🤖. Kaal shows the fellows' work time and productivity at a glance within Discord and a simple CLI tool to track each fellow's coding routine while working for the fellowship. 💪**

— *Gamify the pod's productivity!*

#### Screenshots

###### Registration

![Registration Demo](https://github.com/MLH-Fellowship/pod_1.0.2_kaal/blob/master/branding/register_demo.png)  
  
###### Check In

![Check In](https://github.com/MLH-Fellowship/pod_1.0.2_kaal/blob/master/branding/Screenshot%20from%202020-10-13%2001-12-17.png)

###### Check Out

![Check Out](https://github.com/MLH-Fellowship/pod_1.0.2_kaal/blob/master/branding/Screenshot%20from%202020-10-13%2001-12-26.png)

## 🧑‍🔬 How to use it?

#### Prerequisites:

1. Python3: [Installation Steps](https://phoenixnap.com/kb/how-to-install-python-3-ubuntu)
2. Clone the repository: `https://github.com/MLH-Fellowship/pod_1.0.2_moropy.git`
3. Type `sudo apt-get install xdotool`
4. Head over to the cloned repository's `kaal_cli` folder and type `pip3 install -r requirements.txt`
5. All the dependencies will be installed and the project is good to go! 🎉


#### Instructions:

1. Type `!register` command in the Kaal enabled Discord channel and it'll return your `Secret Code` 🤫 in the DM.
2. Head over to the repository and type `./moropy.py register` and paste the `Secret Code`.
3. When you wish to start the work, type `./moropy.py checkin` and it'll start watching the softwares open. 👀
4. When it's time to stop the work, type `./moropy.py checkout`, the CLI will check you out and stop watching the windows. 😌

## 🙇 Why do we need it?

This project was inspired by [WakaTime](https://wakatime.com/) and the amazing Discord Bots! 

Using this project, the productivity of the pod can be monitored without invading the privacy and can be used to have a healthy competition within the pod. The pod leader (or anyone in the channel) can see the performance of all fellows in that week (or all-time data) in the form of a leaderboard. 🧑🏻‍💻

## 🧑‍💻 Our tech stack!

1. **[Flask](https://flask.palletsprojects.com/en/1.1.x/)** 1️⃣

    Flask is being used as the backend server to listen for updates from CLI, managing database and listening to the discord bot.

2. **[Click](https://pypi.org/project/click/)** 2️⃣

    Click package manages the beautiful CLI and triggers background scripts for listening to window changes.

3. **[discord.py](https://pypi.org/project/discord.py/)**

    `discord.py` helps to provision the REST APIs for the bot-server communication.

4. **[Firestore](https://firebase.google.com/docs/firestore)**

    Firestore is being used as the primary database.

5. **[Heroku](https://www.heroku.com/)**

    Heroku hosts the backend server for the Bot and CLI.

6. **[xdotool](https://www.howtoinstall.me/ubuntu/18-04/xdotool/)**

    xdotool is a linux package to listen to the active window process ID.

## 🙏🏻 Anything else?

Remember the rules!

It's _not_ about:

- how good your code is,
- how novel the idea is, or
- how useful the project is.

It _is_ about:

- **Impressive**: People's wows are on record! Check Discord!
- **Design**: A snappy 24/7 available bot and a beautiful CLI :stars:
- **Completion**: Our hack works completely!
- **Learning**: Discord Bot, Authorization, Bash Scripting and Flask!
- **OSS practices**: Plentiful commits, Code reviews, Kanban boards, issues, branches, PRs!
- **Approved tech**: Flask, Click

In short:

> _Sometimes a pointless project is one of the best hacks!_<br> [—MLH Hackathon Rules](https://github.com/MLH-Fellowship/fellows-0/blob/master/orientation-hackathon/rules.md)


CHANGELOG.md
# Changelog

### Version 0.1.0 [Unreleased]

- _Kaal Discord Bot_ to keep your pod members in the loop
- _Kaal CLI_ brings productivity suite to command line
- _Kaal Backend_ keeps _Kaal Bot_ and _Kaal CLI_ synchronized


moropy_bot/bot.py
import config
import discord
import utils
from discord.ext import commands

client = discord.Client()
bot = commands.Bot(command_prefix='!')

CHANNEL_WEBHOOK_URL = {}


# TODO: Remove code duplicity
@bot.command(name='leaderboard', help='Show leaderboard of the pod')
async def on_leaderboard_message(ctx):
    user = ctx.message.author
    current_channel = ctx.message.channel
    channel_category = current_channel.category.name
    user_roles = [role.name for role in user.roles]
    pod_role = None

    # Check which pod members are to queried
    for role in user_roles:
        if role in channel_category:
            pod_role = role
            break

    status_code, pod_details = utils.get_pod_leaderboard(pod_role)

    if pod_details:
        message = (
            f'<@{user.id}>, following is the current leaderboard of `{pod_role}`\n'
        )
        await current_channel.send(message)
        await current_channel.send(
            '```' + utils.beautify_pod_coding_time(pod_details) + '```'
        )


@bot.command(name='status', help='List availability status of your pod')
async def on_status_message(ctx):
    user = ctx.message.author
    current_channel = ctx.message.channel
    channel_category = current_channel.category.name
    user_roles = [role.name for role in user.roles]
    pod_role = None

    # Check which pod members are to queried
    for role in user_roles:
        if role in channel_category:
            pod_role = role
            break

    status_code, pod_details = utils.get_pod_availability_status(pod_role)

    if pod_details:
        message = (
            f'<@{user.id}>, follwing is the availability of members of `{pod_role}``'
        )
        await current_channel.send(message)
        await current_channel.send(
            '```' + utils.beautify_pod_availability_status(pod_details) + '```'
        )


@bot.command(name='register', help='Start user registration')
async def on_message(ctx):
    user = ctx.message.author

    # Send acknowledgement to user on current channel
    current_channel = ctx.message.channel
    await current_channel.send(
        f'<@{user.id}> we have started registration for you, '
        'keep an eye on your DMs for further instructions. :wink:'
    )

    user_dm = await user.create_dm()
    user_roles = [role.name for role in user.roles]

    # Check whether the user has already registered
    user_exists, userHash = utils.validate_user(user.id)
    if user_exists:
        await current_channel.send(
            f'We have met before <@{user.id}>! I have sent you a present on DM :smiling_imp:'
        )
        await user_dm.send(
            f'There you go with you hashCode: `{userHash}`. '
            'You can now register yourself through the CLI.'
            f'When in doubt, head over to our documentation: {config.DOCUMENTATION_URL}'
        )
        return

    # Send greetings to user on DM
    await user_dm.send(config.LOADING_GIF_URL)
    await user_dm.send(
        f'<@{user.id}> we have started moving things for you! Meanwhile, you '
        'can have a look at our documentation: '
    )

    # Register user on Kaal Backend
    status_code, userHash = utils.registerUser(user.id, user.name, user_roles)

    if userHash:
        await user_dm.send(config.BOOM_GIF_URL)
        await user_dm.send(
            f'Kudos <@{user.id}>! There you go with you hashCode: `{userHash}`. '
            'You can now register yourself through the CLI.'
            'When in doubt, head over to our documentation: '
        )
    else:
        await user_dm.send(
            'Oh no! It seems like we are currently facing a problem registering '
            'you in our database, please report your bug report on our GitHub, '
            f'with this code `{status_code}`. Our developers will help you!'
        )

    await create_webhooks_for_users(ctx, user, userHash)


async def create_webhooks_for_users(ctx, user, userHash):
    # Code for creating webhooks for each channel
    # Get all text_channels grouped by categories
    category_dict = dict(ctx.guild.by_category())

    # Store categories and channels user has access to
    user_categories = []
    user_channels = []
    user_webhooks = []

    # Match category name according to name of the role which user have
    for role in user.roles:
        for category in ctx.guild.categories:
            if role.name in category.name:
                # If the name of a role of user is present in name of a
                # category, then store such category
                user_categories.append(category)

    # For each matched category, store channel in that category
    # which are named 'general'
    # TODO: make channel lookup name configurable by user
    for category in user_categories:
        for channel in category_dict[category]:
            if channel.name == 'general':
                user_channels.append(channel)

    # Create webhooks for all channels in "available channel"
    for channel in user_channels:
        # Try a local lookup for webhook url
        try:
            webhook_url = CHANNEL_WEBHOOK_URL[channel.id]
        except KeyError:
            # Look whether webhook for such URL is available on Kaal backend
            status_code, webhook_url = utils.get_channel_webhook_url(channel.id)
            if not webhook_url:
                webhook = await channel.create_webhook(name='moropy_bot')
                webhook_url = webhook.url
                CHANNEL_WEBHOOK_URL[channel.id] = webhook_url
                utils.create_channel_webhook_url(channel.id, webhook_url)
        finally:
            user_webhooks.append(webhook_url)
    utils.update_users_webhook_url(userHash, user_webhooks)

    for channel in user_channels:
        await channel.send(config.MAKES_SOME_NOISE_GIF_URL)
        await channel.send(f'See who has just joined us, :eyes: <@{user.id}>')


bot.run(config.TOKEN)


moropy_bot/utils.py
from logging import Logger

import requests
from config import KAAL_BACKEND_URL

logger = Logger(__name__)

REGISTER_USER_ENDPOINT = '/register/'
CHANNEL_WEBHOOK_ENDPOINT = '/channel/{}'
USER_WEBHOOKS_ENDPOINT = '/storechannel'
USER_VALIDATE_ENDPOINT = '/validatebot/'
POD_DETAILS_ENDPOINT = '/summary/'


def _get_absolute_url(relative_url):
    return KAAL_BACKEND_URL + relative_url


def registerUser(discord_user_id, discord_username, discord_roles):
    response = requests.post(
        url=_get_absolute_url(REGISTER_USER_ENDPOINT),
        json={
            "userId": discord_user_id,
            "userName": discord_username,
            "roles": discord_roles,
        },
    )
    return (response.status_code, response.json().get('userHash', None))


def get_channel_webhook_url(channel_id):
    response = requests.get(
        url=_get_absolute_url(CHANNEL_WEBHOOK_ENDPOINT.format(channel_id)),
    )
    return response.status_code, response.json().get('userHash', None)


def create_channel_webhook_url(channel_id, webhook_url):
    response = requests.post(
        url=_get_absolute_url(CHANNEL_WEBHOOK_ENDPOINT.format(channel_id)),
        json={'webhook_url': webhook_url},
    )
    return response.status_code, response.json().get('webhook_url', None)


def update_users_webhook_url(userHash, webhook_urls):
    response = requests.post(
        url=_get_absolute_url(USER_WEBHOOKS_ENDPOINT),
        json={'userHash': userHash, 'webhookUrls': webhook_urls},
    )
    status_code = response.status_code
    if response.status_code != 200:
        logger.warn(f'Users webhook URLs update failed, status code: {status_code}')
    return status_code, None


def validate_user(user_id):
    response = requests.post(
        url=_get_absolute_url(USER_VALIDATE_ENDPOINT), json={"userID": str(user_id)}
    )
    if response.status_code == 404:
        return False, None
    return response.json().get('status', False), response.json().get('userHash', None)


def get_pod_availability_status(role):
    response = requests.post(
        url=_get_absolute_url(POD_DETAILS_ENDPOINT), json={"role": str(role)}
    )
    return response.status_code, response.json().get('msg', None)


def get_pod_leaderboard(role):
    response = requests.post(
        url=_get_absolute_url(POD_DETAILS_ENDPOINT), json={"role": str(role)}
    )
    return response.status_code, response.json().get('msg', None)


moropy_cli/ticker.py
#! /usr/bin/python3

import datetime
import json
import os
import pathlib
import subprocess
import time

import requests

# The ticker script creates a moropy.sh file in the tmp folder
# to keep the status of ticker. 0 - ticker should stop,
# 1 - ticker should keep running
# At each second the ticker will check the status from the file
# and then execute the command of getting the active window.

root_dir = pathlib.Path.home()
moropy_dir = os.path.join(root_dir, ".moropy")
status_file_name = "status"
credentials_file_name = "creds.json"
status_file_path = os.path.join(moropy_dir, status_file_name)
credentials_file_path = os.path.join(moropy_dir, credentials_file_name)
flush_file_path = "flush.csv"
base_url = "https://kaal-backend.herokuapp.com"


def read_user_data():

    with open(credentials_file_path, 'r') as file:

        json_object = json.load(file)
        return json_object


def write_to_file(filename, process, start_time):

    if len(process) == 0:
        return

    end_time = datetime.datetime.utcnow()
    duration = (end_time - start_time).total_seconds()

    with open(filename, "a") as window_logs:
        window_logs.write(
            "{},{},{},{}\n".format(previous_window, start_time, end_time, duration)
        )


def push_to_database():

    user_data = read_user_data()
    activities = []
    total_time = 0

    with open("logs.csv", "rb") as file:

        while True:

            line = file.readline()

            if not line:
                break

            process, startTime, endTime, duration = line.decode('UTF-8').split(',')

            total_time += int(float(duration))

            activities.append(
                {
                    "name": process,
                    "start_time": startTime,
                    "end_time": endTime,
                    "duration": int(float(duration)),
                }
            )

    payload = {
        "userHash": user_data['userHash'],
        "activities": activities,
        "codingTime": total_time,
    }

    payload_json = json.dumps(payload)

    # print(payload_json)

    response = requests.post("{}/storeactivity/".format(base_url), payload_json)

    print(response.status_code)

    # with open(flush_file_path, "w") as f:
    # f.write(payload_json)

    with open("logs.csv", "w") as file:
        pass


# writes active status to moropy.sh
with open(status_file_path, 'w') as file:
    file.write("1")

ticker_continue = True

previous_window = ""
previos_start_time = datetime.datetime.utcnow()
previos_push_time = datetime.datetime.utcnow()

with open("logs.csv", "w") as window_logs:
    pass

while ticker_continue:

    # reads the status from moropy.sh
    with open(status_file_path, 'rb') as file:
        script = file.readlines()
        ticker_continue = bool(int(script[0].decode('UTF-8')))

    if not ticker_continue:
        write_to_file("logs.csv", previous_window, previos_start_time)
        push_to_database()
        break

    if (datetime.datetime.utcnow() - previos_push_time).total_seconds() >= 1 * 60:
        push_to_database()
        previos_push_time = datetime.datetime.utcnow()

    # if ticker_continue is false, ticker will stop after this iteration

    cmd = "xdotool getwindowfocus getwindowname getwindowpid"

    output = subprocess.getoutput(cmd)

    pid = output.split('\n')[1]

    process_name_cmd = "ps -p {} -o comm=".format(pid)

    status, process_name = subprocess.getstatusoutput(process_name_cmd)

    if status == 0 and process_name != previous_window:

        with open("logs.csv", "a") as window_logs:

            write_to_file("logs.csv", previous_window, previos_start_time)

            previous_window = process_name
            previos_start_time = datetime.datetime.utcnow()

    time.sleep(0.5)


moropy_backend/app.py
import requests as request_lib
from firebase_services import (
    get_user,
    getChannel,
    makeChannel,
    store_activity,
    update,
    updateWebhooks,
    upload,
)
from flask import Flask, jsonify, request

app = Flask(__name__)


@app.route('/', methods=['GET'])
def welcome():
    return jsonify({"status": "The API is healthy and working properly"})


@app.route('/register/', methods=['POST'])
def register_user():
    input_json = request.get_json(force=True)
    userHash = upload(input_json['userId'], input_json['roles'], input_json['userName'])
    return jsonify({"statusCode": 200, "userHash": userHash})


@app.route('/validate/', methods=['POST'])
def get_user_details():
    input_json = request.get_json(force=True)
    resp = get_user(input_json['userHash'])
    if resp == 'User not found':
        return jsonify({'msg': "User does not exist"}), 401
    else:
        return jsonify({'user': resp}), 200


@app.route('/storeactivity/', methods=['GET', 'POST'])
def activity():
    if request.method == 'GET':
        return jsonify({"msg": "This is a POST route"})
    else:
        input_json = request.get_json(force=True)
        ret = store_activity(input_json['userHash'], input_json['activities'])
        if ret == 'working':
            return jsonify({"msg": "Activity stored successfully"}), 200
        else:
            return jsonify({"msg": "Error"}), 400


@app.route('/status/', methods=['GET', 'POST'])
def update_status():
    if request.method == 'GET':
        return jsonify({"msg": "This is a POST route"})
    else:
        input_json = request.get_json(force=True)
        # Update status in database
        user_hash = input_json['userHash']
        user_status = input_json['status']
        ret = update(user_hash, user_status)
        if ret == 'successful':
            notify_on_discord(user_hash, user_status)
            return jsonify({"msg": "Status updated successfully"}), 200
        else:
            return jsonify({"msg": "Error"}), 400


@app.route('/storechannel', methods=['POST'])
def update_web():
    input_json = request.get_json(force=True)
    ret = updateWebhooks(input_json['userHash'], input_json['webhookUrls'])
    if ret is True:
        return jsonify({"msg": "webhooks stored successfully"}), 200
    else:
        return jsonify({"msg": "Error"}), 400


@app.route('/channel/<string:channel_id>', methods=['GET', 'POST'])
def get_channel_details(channel_id):
    if request.method == 'GET':
        ret = getChannel(channel_id)
        if ret != '':
            return jsonify({"channel": ret}), 200
        else:
            return jsonify({"msg": "Error"}), 400
    else:
        input_json = request.get_json(force=True)
        ret = makeChannel(channel_id, input_json['webhook_url'])
        if ret is True:
            return jsonify({"msg": "Channel created successfully"}), 200
        else:
            return jsonify({"msg": "Error"}), 400


def notify_on_discord(user_hash, status):
    user = get_user(user_hash)
    user_channel_urls = user['webhookUrls']
    username = user['userName']
    if status == 'Away':
        message = (
            f'\U0001F534 `{username}` has checked out! '
            'Friendly reminder for everyone to stretch.'
        )
    else:
        message = f'\U0001F7E2 `{username}` has just checked-in, good day `{username}`!'

    for url in user_channel_urls:
        request_lib.post(url, json={'content': message, 'tts': False})


if __name__ == '__main__':
    app.run(host="0.0.0.0", port=1500, debug=True)


moropy_backend/firebase_services.py
import uuid

import firebase_admin
from firebase_admin import credentials, firestore

# Use a service account
cred = credentials.Certificate('firebase-key.json')
firebase_admin.initialize_app(cred)

db = firestore.client()
batch = db.batch()


def upload(userID, roles, userName):
    user_ref = db.collection(u'users')

    userHash = uuid.uuid4()
    doc_ref = user_ref.document(f"{userHash}")
    doc_ref.set(
        {
            u'roles': roles,
            u'discordId': str(userID),
            u'userName': userName,
            u'status': 'Away',
        }
    )
    return userHash


def get_user(userHash):
    user_ref = db.collection(u'users')

    users = user_ref.stream()
    for user in users:
        if user.id == userHash:
            return user.to_dict()
    return 'User not found'


def validate_user(userID):
    duplicates = db.collection(u'users').where(u'discordId', u'==', (userID)).stream()

    print('here ' + userID)
    # duplicates = user_ref.stream()
    for duplicate in duplicates:
        print(duplicate.to_dict()['discordId'])
        if str(userID) == str(duplicate.to_dict()['discordId']):
            return duplicate.id
    return 'User not found'


def store_activity(userHash, activity):
    user_ref = db.collection(u'users')

    doc_ref = user_ref.document(f'{userHash}').collection('activity')
    for i in range(len(activity)):
        doc_ref.document().set(activity[i])
    return "working"


def update(userHash, status):
    user_ref = db.collection(u'users')

    doc_ref = user_ref.document(f'{userHash}')
    doc_ref.update(
        {
            u'status': status,
        }
    )
    return 'successful'


def updateWebhooks(userHash, webhookUrls):
    user_ref = db.collection(u'users')

    doc_ref = user_ref.document(f'{userHash}')
    doc_ref.update({u'webhookUrls': webhookUrls})
    return True


def makeChannel(channel_id, url):
    channel_ref = db.collection(u'channels')
    channel_ref.document(channel_id).set({"url": url})
    return True


def getChannel(channel_id):

    channel_ref = db.collection(u'channels')
    res = channel_ref.stream()
    for r in res:
        if r.id == channel_id:
            print(True)
            return r.to_dict()

    return ''


moropy_cli/moropy.py
#! ./venv/bin/python3

import datetime
import json
import os
import pathlib
import subprocess

import click
import requests

root_dir = ""
credentials_file_name = "creds.json"
status_file_name = "status"
base_url = "https://kaal-backend.herokuapp.com"


def read_user_data():

    credentials_file_path = os.path.join(root_dir, credentials_file_name)

    with open(credentials_file_path, 'r') as file:

        json_object = json.load(file)
        return json_object


@click.command('register')
def register_user():

    click.echo("👋 Welcome to Kaal!")
    click.echo()

    code = click.prompt("🤫 Enter your secret code here", type=str)
    click.echo()

    click.echo("⏱️  Please wait while I validate your code!")
    click.echo()

    credentials_file_path = os.path.join(root_dir, credentials_file_name)

    # section to contact the backend API to validate code

    response = requests.post(
        "{}/validate".format(base_url), json.dumps({"userHash": code})
    )

    if response.status_code == 401:
        print("🤨 I found your token to be invalid. Please try again!")
        return

    if response.status_code != 200:
        print(
            "🤯 I couldn't reach the Kaal servers. Please try again or contact the admin."
        )
        return

    response_data = response.json()

    user_data = {
        "userHash": code,
        "userName": response_data['user']['userName'],
    }

    click.secho("✅ Congratulations ", nl=False)
    click.echo(
        click.style("{}".format(user_data['userName']), bold=True, fg='yellow'),
        nl=False,
    )
    click.echo("! Registration process is successful.\n")
    click.echo("🥺 I can't wait to hear the checkin and checkout commands.")

    file_data = json.dumps(user_data, indent=4)

    with open(credentials_file_path, "w") as file:
        file.write(file_data)


@click.command('checkin')
def checkin():

    user_data = read_user_data()

    click.echo()

    click.echo(
        click.style("🌞 Checking you in {}!".format(user_data['userName']), fg='green')
    )

    hours = datetime.datetime.now().time().hour
    minutes = datetime.datetime.now().time().minute

    click.echo()

    click.echo("⏲  The time is ", nl=False)
    click.secho("{}:{}".format(hours, minutes), bold=True)

    click.echo()

    click.echo("🧐 Getting some popcorn! 🍿 It's interesting to watch you work!")

    subprocess.call("chmod +x ./ticker.py", shell=True)
    subprocess.call("nohup ./ticker.py &", shell=True)
    # subprocess.call("nohup ./ticker.py >/dev/null 2>&1 &", shell=True)


@click.command('checkout')
def checkout():
    status_file_path = os.path.join(root_dir, status_file_name)

    user_data = read_user_data()

    click.echo()

    click.echo(
        click.style(
            "🌞 Checking you out {}!".format(user_data['userName']), fg='bright_black'
        )
    )

    hours = datetime.datetime.now().time().hour
    minutes = datetime.datetime.now().time().minute

    click.echo()

    click.echo("⏲  The time is ", nl=False)
    click.secho("{}:{}".format(hours, minutes), bold=True)

    click.echo()

    if not os.path.exists(status_file_path):
        print(status_file_path, "doesn't exist")
        return

    # writes close status to moropy.sh
    with open(status_file_path, 'w') as file:
        file.write("0")

    click.echo("😌 Great work today! Going to sleep! Bye! 🙌")


@click.command('away')
def set_away():

    click.echo("Setting away")

    credentials_file_path = os.path.join(root_dir, credentials_file_name)

    with open(credentials_file_path, "rb") as file:
        user_hash = file.readline().decode('UTF-8')[:-1]
        # user_name = file.readline().decode('UTF-8')

        data = {'userHash': user_hash, 'status': 'away'}

        res = requests.post("{}/status/".format(base_url), json.dumps(data))

        print(res.json())


@click.command('available')
def set_available():

    click.echo("Setting available")

    credentials_file_path = os.path.join(root_dir, credentials_file_name)

    with open(credentials_file_path, "rb") as file:
        user_hash = file.readline().decode('UTF-8')[:-1]
        # user_name = file.readline().decode('UTF-8')

        data = {'userHash': user_hash, 'status': 'available'}

        res = requests.post("{}/status/".format(base_url), json.dumps(data))

        print(res.json())


@click.group()
def init_cli():

    user_home_dir = pathlib.Path.home()

    global root_dir
    root_dir = pathlib.Path.joinpath(user_home_dir, ".moropy")

    if not os.path.exists(root_dir):
        os.mkdir(root_dir)
        click.echo("generated {}".format(root_dir))


init_cli.add_command(register_user)
init_cli.add_command(checkin)
init_cli.add_command(checkout)
init_cli.add_command(set_away)
init_cli.add_command(set_available)

if __name__ == '__main__':
    init_cli()
